File: swf-tree.ts/src/lib/avm1/value.ts
import { TaggedUnionType } from "kryo/types/tagged-union";
import * as values from "./values/index";

export type Value =
  values.Boolean
  | values.String
  | values.Constant
  | values.Float32
  | values.Float64
  | values.Sint32
  | values.Null
  | values.Register
  | values.Undefined;

export const $Value: TaggedUnionType<Value> = new TaggedUnionType<Value>(() => ({
  variants: [
    values.$Boolean,
    values.$String,
    values.$Constant,
    values.$Float32,
    values.$Float64,
    values.$Sint32,
    values.$Null,
    values.$Register,
    values.$Undefined,
  ],
  tag: "type",
}));

File: swf-tree.rs/src/avm1/value.rs
use float_is::Is;

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "type", content = "value", rename_all = "kebab-case")]
pub enum Value {
  String(String),
  Float32(f32),
  Null,
  Undefined,
  Register(u8),
  Boolean(bool),
  Float64(f64),
  Sint32(i32),
  Constant(u16),
}

impl ::std::cmp::PartialEq for Value {
  fn eq(&self, other: &Self) -> bool {
    match (self, other) {
      (&Value::Float32(left), &Value::Float32(right)) => left.is(&right),
      (&Value::Float64(left), &Value::Float64(right)) => left.is(&right),
      (left, right) => left == right,
    }
  }

  fn ne(&self, other: &Self) -> bool {
    !self.eq(other)
  }
}

impl ::std::cmp::Eq for Value {}
